Load backends in a lazy way

datapythonista opened the pull request:
Closes #2827 

@jreback, there is a problem here, not sure how to move forward with.

Currently, pyarrow data types are registered into Ibis in the parquet backend. Since now backends are loaded in a lazy way, the next code fails unless the parquet backend is loaded first:
```python
import pyarrow
import ibis.expr.datatypes as dt

dt.dtype(pyarrow.int8())
```
There is a test failing because of this. I can easily solve it by calling `ibis.parquet` before the test. But not sure if that's what we want. Do we really want to register arrow types when the parquet backend is loaded? Or when?

jreback commented:
> Do we really want to register arrow types when the parquet backend is loaded? Or when?

right this is a fundamental issue. in theory we would want to actually load types w/o the entire backend, but this is not easily possible. So really no choice but to load when the backend is loaded. 

where is this a problem exactly?

datapythonista commented:
> where is this a problem exactly?

The test of the pyarrow dtypes is an Ibis core test, and the registration is in the parquet backend. So, the problem is when testing backends without the parquet backend being loaded. Those are the tests: `ibis/tests/expr/test_datatypes_arrow.py::test_convert_arrow`

For now I'm just loading the pandas and parquet backends at Ibis start. I think we need changes to the architecture of Ibis to solve properly. I opened #2834 for now, but I think there is more than that.

jreback commented:
> For now I'm just loading the pandas and parquet backends at Ibis start. I think we need changes to the architecture of Ibis to solve properly. I opened #2834 for now, but I think there is more than that.

yep sounds like a good idea to investigate

jreback commented on ibis/config.py:
```diff
@@ -200,7 +202,17 @@ def __getattr__(self, key):
-            raise AttributeError(*e.args)
+            # See if the options doesn't exist because it's from a backend
+            # not yet loaded. Try to load, and try the option again.
+            try:
```
do we go down this path in tests?

datapythonista replied:
Good point. Looks like we don't have any tests for options. Our options code is a copy of the pandas one, which does have tests. I don't think copying pandas tests, or implementing new ones are great ideas. If we want to do things properly, we should create a new package for the options code, and use it in both projects.

Let me know your thoughts, but in any case, I don't think this is for this PR.

jreback replied:
i think you can remove *all* of this because you *already* load the options in the backend loading itself. I woudn't load the backend here at all.

datapythonista replied:
Unfortunately this is needed. See this code:
```python
ibis.options.impala. temp_table = 'foo'
ibis.impala.connect(...)
```
Without this code, the first line will fail, since there is no `impala.temp_table` option registered. The loading of the backend you're referring to is loaded in the second line.

jreback replied:
this is easy to replace by
```
import ibis.impala
ibis.options.impala.....
ibis.impala.connect()
```
right?

datapythonista replied:
It would be:
```python
import ibis
ibis.impala
ibis.options.impala.temp_table = 'foo'
ibis.impala.connect()
```
But, yes, that's correct, if we are happy that the next code fails when setting the option, this is not needed:
```python
import ibis
ibis.options.impala.temp_table = 'foo'
ibis.impala.connect()
```

jreback replied:
see comments above

jreback replied:
i don't really love all of these nested try/excepts

jreback commented:
prob rebase to fix dask

datapythonista commented:
@jreback this is ready now. Main changes since you last reviewed:
- We register backends options on `ibis.__getattr__`, so in tests I had to replace `ibis.backends.dask.Backend()` by `ibis.dask`, so options are registered.
- Looks like the Spark backend got the `FloorDivide` operation registered from the PySpark backend. Now that PySpark is not necessarily loaded, had to implement it there.

Thanks!

jreback commented on ibis/__init__.py:
```diff
@@ -51,15 +53,29 @@
-            backend.register_options()
+
+def __getattr__(name: str):
+    entry_points = list(
```
can you add a doc-string and return type

jreback commented on ibis/__init__.py:
```diff
@@ -51,15 +53,29 @@
+    )
+    if len(entry_points) == 0:
+        raise AttributeError(
+            f"module 'ibis' has no attribute '{name}'. "
```
are these 2 tested? why are we warning on multiple entry points? (and not raising)

jreback commented:
do you need to update any docs?

icexelloss commented on ibis/__init__.py:
```diff
@@ -1,12 +1,14 @@
+
 import pkg_resources
 
+# Converting an Ibis schema to a pandas DataFrame requires registering
```
Sorry, do you mean `converting an Ibis schema to a pandas dtype` here? I am not sure what do you mean by converting ibis schema to a pandas DataFrame..

icexelloss replied:
Also IMO, I think it's OK to treat pandas backend special since pandas can be a hard dependency and is kind of baked in the ibis core/expr layer.

datapythonista replied:
Thanks, yes, this is not clear, will rephrase. This refers to the `Schema.apply_to` I explained below.

Personally I don't think pandas should be a hard dependency, and the pandas backend shouldn't be preloaded. I don't even think the pandas backend should be shipped with Ibis in the long term. You can see my proposed roadmap at #2828, so you can have an idea on my thoughts for the project. And any feedback there is very welcome.

jreback replied:
but please do not change at this time. pandas is not a big deal as a dep at all.

datapythonista replied:
Mhh, the change is not to make pandas not be a dependency (even if it would help if we do that). The change is trivial, just changing the function to the obvious file where it should be. And also makes it a regular method, instead of injecting it as a method in a hacky way. I think it's a small, but useful clean up, with no drawbacks. What is the problem with keeping it?

icexelloss commented on ibis/__init__.py:
```diff
@@ -51,15 +53,29 @@
+        )
+    elif len(entry_points) > 1:
+        warnings.warn(
+            f"More than one entrypoint found for backend '{name}', "
```
Sorry not super familiar with this. What exactly mean to have more than one entrypoint for backend?

icexelloss replied:
(And why do we want to warn on this?)

datapythonista replied:
Imagine that two different developers create packages for bigquery. And both of them add an entrypoint with the name `bigquery`. When a user with both packages installed call `ibis.bigquery`, both will be found, but the first to be found is used, and the second will be ignored. @jreback suggests to raise an exception. I thought it'd make sense to move forward, knowing the user (with this warning) which of the packages will be used.

I'm ok raising an exception. And for your feedback, I think it makes sense to rephrase the message, to make it less technical, and more user friendly. Thanks for it.

jreback replied:
let's raise an exception. this is not tested at all in any event.

datapythonista replied:
I changed it, now it's raising an exception, and there is a test for it.

icexelloss commented on ibis/backends/spark/compiler.py:
```diff
@@ -50,6 +50,12 @@ class SparkExprTranslator(ExprTranslator):
 rewrites = SparkExprTranslator.rewrites
 
 
+@rewrites(ops.FloorDivide)
```
I like this.

But..is this related to this PR? XD

datapythonista replied:
Somehow related. This is missing in master and tests fail because the lack of registering `FloorDivide` in Spark. In master tests don't really fail, because the PySpark backend is also loaded, and looks like the PySpark backend is also registering the Spark operation. I need to change this here (or in a separate PR merged before this one), because here we load one backend at a time, and without loading PySpark, the Spark backend tests fail.

I'm already working on removing the spark backend, so this addition makes the tests pass here, but this code won't live very long, so no big deal I think, even an empty function would be ok IMO.

icexelloss replied:
kk thanks for the explanation. SGTM

icexelloss commented on ibis/config.py:
```diff
@@ -200,7 +202,17 @@ def __getattr__(self, key):
             v = self.d[key]
         except KeyError as e:
-            raise AttributeError(*e.args)
+            # See if the options doesn't exist because it's from a backend
```
In what way it breaks if we didn't do this here? It seems to me to be a reasonable behavior to error out if user tries to access an option defined by a backend without importing that backend first...

datapythonista replied:
This change makes this code work:
```python
import ibis
ibis.options.impala.temp_table = 'foo'
ibis.impala.connect(...)
```
If we break here, and make users write the next code, then wouldn't be needed (some other changes would be needed to this PR for it to work):
```python
import ibis
import ibis.backends.impala

ibis.options.impala.temp_table = 'foo'
ibis.impala.connect(...)
```

jreback replied:
yes i think thi sis better

icexelloss replied:
kk SGTM

datapythonista replied:
Should we simply use something like:
```python
import ibis.backends.impala

ibis.backends.impala.set_option('temp_table', 'foo')
ibis.backends.impala.connect(...)
```
or the syntax I proposed in the past
```python
import ibis

ibis.backends.impala.connect(..., temp_table='foo')
# or
ibis.backends.impala.connect(..., options={'temp_table': 'foo'})
```


I think `ibis.<backend-name>` is already cumbersome, and caused several problems. And IMHO this is becoming even more cumbersome, to have to import an unrelated module to make the above code work is far from what I'd expect to have to write as a user.

jreback replied:
i prefer

```
import ibis.backends.impala 
ibis.backends.impala.set_option('temp_table', 'foo') 
ibis.backends.impala.connect(...)```
```

connect has nothing to do with setting the options of the backend itself (and may actually have multiple invokations). typing these together is not great.

all that said this can be a followup.

icexelloss commented on ibis/expr/schema.py:
```diff
@@ -9,6 +9,28 @@
+
+Returns
+-------
+result : pd.Series
```
A bit confused.. Does return a dtype or a pd.Series?

datapythonista replied:
I just moved this function of file without change, probably better to fix things in a follow up. I can have a look since this is just the docs, but in general better to not make changes that don't show in a commit diff.

icexelloss replied:
kk SGTM

icexelloss commented on ibis/expr/schema.py:
```diff
@@ -133,6 +155,39 @@ def name_at_position(self, i):
         return self.names[i]
 
+    def apply_to(self, df):
+        """Applies the Ibis schema to a pandas DataFrame
```
Sorry what does this method do? Does it cast the types of df according to this ibis schema?

datapythonista replied:
As said in the previous comment, I just moved this method, since IMHO it was in the wrong file. And I finally ended up always loading the pandas backend, but if we eventually don't load it at `import ibis`, this change would be needed, since this code is needed even if the backend is not loaded.

But answering your question, my understanding is that when we convert the cursor returned by the backend to pandas, types are inferred by pandas. And this method takes the original schema of the ibis expression (or let's say the SQL table), and casts columns to have the same types.

icexelloss replied:
kk SGTM

icexelloss commented:
@datapythonista Looks good high level! I left some comments and would appreciate it if you take a look.

datapythonista commented:
I think I addressed all comments, and commented in the ones were I think no change should be make. Please let me know if I missed anything, or if you have further comments.

icexelloss commented:
Thanks @datapythonista. +1 from me.

jreback commented:
ok happy to merge this, but i do thing before release should fix up the option setting stuff.

datapythonista commented:
> ok happy to merge this, but i do thing before release should fix up the option setting stuff.

Thanks. I guess we may want to keep backward compatibility for Ibis 2.0, with deprecation warnings where needed. So, I guess it may make sense to keep this code until after the release. But next week I'll be opening issues to discuss more in detail what we exactly want to have, for the release and long term.
